### PR TITLE
loss averaging across multiple devices

### PR DESCRIPTION
**Pull request by hadipash**
## Motivation

Average loss across devices to see better training stats.

### CODE REVIEW

File: mindocr/utils/callbacks.py [hunk 1]
```diff
@@ -158,30 +163,42 @@ def __init__(self,
...
         self.step_start_time = time.time()
 
+        self._losses = []
+        if device_num is not None:
+            reduce_sum = ms.ops.AllReduce()
+
+            @jit
+            def reduce(x):                          # lamda expression is not supported in MindSpore
+                return reduce_sum(x) / device_num   # average loss across all cards
+
```
Comment (zhtmike) [inline mindocr/utils/callbacks.py:197]: it is better to put reduce outside the function

**Review by zhtmike — APPROVED** [review]

File: mindocr/utils/callbacks.py (hunk 1 above)
Comment (hadipash) [reply mindocr/utils/callbacks.py:197]: Talked to Jun about it, he said that: 1. callbacks are always executed in native mode, 2. `ops.AllReduce()` may take a noticeable amount of time in native mode due to some overhead computations, so it is generally recommended to wrap it with `jit`.

Although, I agree that reducing single number tensors may be very quick and `jit` could be an overkill here. Maybe we can benchmark later and see if it is really necessary.

File: mindocr/utils/callbacks.py [hunk 2]
```diff
@@ -208,16 +208,15 @@ def __init__(self,
...
+        self._reduce_sum = ms.ops.AllReduce()
+        self._device_num = device_num
+        # lamda expression is not supported in jit
+        self._loss_reduce = self._reduce if device_num is not None else lambda x: x
 
-            self._loss_reduce = reduce
-        else:
-            self._loss_reduce = lambda x: x
+    @jit
+    def _reduce(self, x):
```
Comment (hadipash) [inline mindocr/utils/callbacks.py:218]: @zhtmike @SamitHuang Please check.

**Review by zhtmike — APPROVED** [review]

File: mindocr/utils/callbacks.py (hunk 2 above)
Comment (SamitHuang) [reply mindocr/utils/callbacks.py:218]: Not sure whether running with ms_function in callback is a stable choice. For MS 1.9, pynative with ms_function is not as stable as MS 2.0. If using ms_function is risky, i don't it is worthy to add jit/ms_function considering the ignorable acceleration on this  one-step division computation.

**Review by SamitHuang — APPROVED** [review]

**Pull request merged**